src/coreclr/nativeaot/System.Private.CoreLib/src/System/Threading/Lock.NativeAot.cs

```diff
@@ -92,108 +81,83 @@ internal void Reenter(uint previousRecursionCount)
-                    {
-                        if (sleep)
-                        {
-                            Thread.UninterruptibleSleep0();
-                        }
-                        else
-                        {
-                            Thread.SpinWait(1);
-                        }
-
-                        stage = (StaticsInitializationStage)Volatile.Read(ref s_staticsInitializationStage);
-                        if (stage == StaticsInitializationStage.Complete)
-                        {
-                            goto case StaticsInitializationStage.Complete;
-                        }
-                        else if (stage == StaticsInitializationStage.NotStarted)
-                        {
-                            goto default;
-                        }
-
-                        if (State.TryLock(this))
-                        {
-                            return TryLockResult.Locked;
-                        }
-
-                        sleep = !sleep;
-                    }
-
-                default:
-                    Debug.Assert(stage == StaticsInitializationStage.NotStarted);
-                    if (TryInitializeStatics())
-                    {
-                        goto case StaticsInitializationStage.Complete;
-                    }
-                    goto case StaticsInitializationStage.Started;
+                if (s_staticsInitializationStage == (int)StaticsInitializationStage.NotStarted &&
+                    Interlocked.CompareExchange(
+                        ref s_staticsInitializationStage,
+                        (int)StaticsInitializationStage.Started,
+                        (int)StaticsInitializationStage.NotStarted) == (int)StaticsInitializationStage.NotStarted)
+                {
+                    ScheduleStaticsInit();
+                }
             }
         }
 
-        [MethodImpl(MethodImplOptions.NoInlining)]
-        private static bool TryInitializeStatics()
+        internal static void ScheduleStaticsInit()
         {
-            // Since Lock is used to synchronize class construction, and some of the statics initialization may involve class
-            // construction, update the stage first to avoid infinite recursion
-            switch (
-                (StaticsInitializationStage)
-                Interlocked.CompareExchange(
-                    ref s_staticsInitializationStage,
-                    (int)StaticsInitializationStage.Started,
-                    (int)StaticsInitializationStage.NotStarted))
+            // initialize essentials
+            // this is safe to do as these do not need to take locks
+            s_maxSpinCount = DefaultMaxSpinCount << SpinCountScaleShift;
+            s_minSpinCount = DefaultMinSpinCount << SpinCountScaleShift;
+
+            // we can now use the slow path of the lock.
+            Volatile.Write(ref s_staticsInitializationStage, (int)StaticsInitializationStage.Usable);
```

Comment:
The lock is now functional and did not do anything that could take locks. The rest of initialization is optional and just need to eventually happen.